### ISSUE TITLE
RuleSlot exact rule identification

### ISSUE THREAD

**Issue by DJManuel** [rtucek/vue-query-builder#81]
Hi, first of all thank you for such a great library!

I am facing an issue where I have multiple rules using the same component.
I noticed this behavior when trying to get data into the (custom) component of a (custom / scoped) RuleSlot. Within this component, I have a drop-down that eventually will fetch data from a DB / API. In the end there will be ~30 rules that have a dropdown selection, and in my opinion it is not worth creating 30 custom (pretty much duplicate) dropdown components. Therefor I am trying to use a single component in multiple rules, with different data fetched in the dropdown.

One of the main issue is the identification of the exact rule based on the data that is being forwarded into the RuleSlot:
[src/QueryBuilderRule.vue#L81](../blob/main/src/QueryBuilderRule.vue#L81) and [src/QueryBuilderRule.vue#L42](../blob/main/src/QueryBuilderRule.vue#L42)
Only with this information currently provided, the exact rule can only be found out by matching the `ruleComponent` (and `ruleData`, dependend if `initialValue` is different): 
```
    ruleName() {
      return this.rules
        .find(r => r.component === this.ruleCtrl.ruleComponent)
        .name;
    },
```
however, as I have multiple rules that have the same component, this won't work (it only give back the first element in the rules array it finds).

To fix this behavior, the `identifier` from the rules would be needed inside the RuleSlot.

Basically in the `get ruleSlotProps` ([src/QueryBuilderRule.vue#L42](../blob/main/src/QueryBuilderRule.vue#L42)) adding one item to the object being forwarded to the RuleSlot:
```
  get ruleSlotProps(): RuleSlotProps {
    return {
      ruleComponent: this.component,
      ruleData: this.query.value,
      ruleIdentifier: this.query.identifier
      updateRuleData: (ruleData: any) => this.ruleUpdate(ruleData),
    };
  }
```

When this `ruleIdentifier` is available in the RuleSlot the ruleName (and exact rule) can be found:
```
.find(r => r.identifier=== this.ruleCtrl.ruleIdentifier)
```

Thank you!

**Comment by rtucek** [comment]
Hi @DJManuel and thank you very much for your feedback.

One thing I can't follow is this statement:

> To fix this behavior, **the identifier from the rules would be needed inside the RuleSlot**.

How, would the component's identifier help you in the RuleSlot context to avoid having several components of the same type?
Assume you'd have already support for it today, can you give me some sample code of how you'd make use of it for solving your problem?

Pertaining the problem you're trying to solve:

> I am facing an issue where I have multiple rules using the same component.
I noticed this behavior when trying to get data into the (custom) component of a (custom / scoped) RuleSlot. Within this component, I have a drop-down that eventually will fetch data from a DB / API. In the end there will be ~30 rules that have a dropdown selection, and in my opinion it is not worth creating 30 custom (pretty much duplicate) dropdown components. Therefor I am trying to use a single component in multiple rules, with different data fetched in the dropdown.

You may want to check #47 which seem to be related. I think [my suggestion with a concrete code sample](https://github.com/rtucek/vue-query-builder/issues/47#issuecomment-855473509) should help - it basically extends a dropdown component. Is that what you're looking for?

**Comment by rtucek** [comment]
Have you looked at this code sandbox sample I referred you to in the linked thread above?

Here's the link:
[![Edit Dynamic dropdown](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/dynamic-dropdown-1c48z?fontsize=14&hidenavigation=1&module=%2Fsrc%2FApp.vue&theme=dark)

The only difference is that you have to declare some logic _from where_ the data shall be passed onto your component. In this sample, I demonstrate it with individual AJAX requests via axios upon the component's mount event. Note that the `Dropdown.vue` gets dynamically extended with a dedicated "data factory".

That said, all you'd want to do is to pass a slightly different for logic for...
- ...querying your database
- ...inject data that you've already at hand via Vuex or prop

If it doesn't solve your problem, please let me know why :wink:

**Comment by DJManuel** [comment]
Yes, I have had a look at it but there are a few issues that I found with this implementation.
The code duplication with more than 30 different dropdown components is quite large, in the implementation this adds 10 lines and this results in total to 300 additional lines that are completely duplicate.

Also, I have tried to implement it and it didn't work as expected.
The component doesn't load properly without any log output on what fails.

I have created a MR #82 for this change to be implemented.

Thank you!

**Comment by JhumanJ** [comment]
Hey there, thanks for the great work. I have the exact same issue, and although your "extend" trick can do the job, I think it would make more sense to have the option to directly pass props to the components. Maybe something like that:
```
 {
          identifier: property.nf_id,
          name: property.name,
          component: ColumnFilter,
          initialValue: null,
          componentProps: {
               //....
          }
        }
```
Having the identifier also works but then again it required additional work. What do you guys think? I'm happy to collaborate on this

**Comment by rtucek** [comment]
I've just merged #83 and released it with [v1.2.0](https://github.com/rtucek/vue-query-builder/releases/tag/v1.2.0).

Kudos to @DJManuel for his contribution :tada:
